spyder_kernels/customize/code_runner.py

```diff
@@ -0,0 +1,593 @@
+                filename=filename,
+                canonic_filename=canonic_filename,
+                exec_fun=debug_exec,
+                post_mortem=post_mortem,
+                context_globals=self.shell.user_ns,
+                context_locals=local_ns,
+            )
+
+    def profilecell(self, cellname, filename=None, post_mortem=False):
+        """
+        Profile a code cell from an editor.
+        """
+        local_ns = self.shell.get_local_scope(1)
+        if filename is None:
+            filename = self._get_current_file_name()
+        canonic_filename = canonic(filename)
+
+        with self._profile_exec() as prof_exec:
+            return self._exec_cell(
+                cell_id=cellname,
+                filename=filename,
+                canonic_filename=canonic_filename,
+                exec_fun=prof_exec,
+                post_mortem=post_mortem,
+                context_globals=self.shell.user_ns,
+                context_locals=local_ns,
+            )
+
+    @no_var_expand
+    @needs_local_scope
+    @line_cell_magic
+    def profile(self, line, cell=None, local_ns=None):
+        """Profile the given line."""
+        if cell is not None:
+            line += "\n" + cell
+        with self._profile_exec() as prof_exec:
+            return prof_exec(line, self.shell.user_ns, local_ns)
+
+    @contextmanager
+    def _debugger_exec(self, filename, continue_if_has_breakpoints):
+        """Get an exec function to use for debugging."""
+        if not self.shell.is_debugging():
+            debugger = SpyderPdb()
+            debugger.set_remote_filename(filename)
+            debugger.continue_if_has_breakpoints = continue_if_has_breakpoints
+            yield debugger.run
+            return
+
+        session = self.shell.pdb_session
+        sys.settrace(None)
+        # Create child debugger
+        debugger = SpyderPdb(
+            completekey=session.completekey,
+            stdin=session.stdin, stdout=session.stdout)
+        debugger.use_rawinput = session.use_rawinput
+        debugger.prompt = "(%s) " % session.prompt.strip()
+
+        debugger.set_remote_filename(filename)
+        debugger.continue_if_has_breakpoints = continue_if_has_breakpoints
+
+        try:
+            def debug_exec(code, glob, loc):
+                return sys.call_tracing(debugger.run, (code, glob, loc))
+            # Enter recursive debugger
+            yield debug_exec
+        finally:
+            # Reset parent debugger
+            sys.settrace(session.trace_dispatch)
+            session.lastcmd = debugger.lastcmd
+
+            # Reset _previous_step so that get_pdb_state() notifies Spyder about
+            # a changed debugger position. The reset is required because the
+            # recursive debugger might change the position, but the parent
+            # debugger (self) is not aware of this.
+            session._previous_step = None
+
+    @contextmanager
+    def _profile_exec(self):
```

Comment:
get a profiler that maybe works in a debugger

Reply:
This should also be moved to the other PR I mentioned.